src/crypto.rs

```diff
@@ -0,0 +1,206 @@
+    }
+    /// Returns the GCM tag property.
+    fn get_property_gcm_tag(&self) -> Result<Vec<u8>>
+    {
+        context_get_property_multiple(self, types::Property::GcmTag)
+    }
+    /// Returns the CCM tag property.
+    fn get_property_ccm_tag(&self) -> Result<Vec<u8>>
+    {
+        context_get_property_multiple(self, types::Property::CcmTag)
+    }
+}
+
+/// Implementation of all GCM/CCM properties for decrypt/open
+pub trait ContextWithXcmDecryptProperties: Context {
+    /// Sets the GCM AAD property.
+    fn set_property_gcm_aad(&self, gcm_aad: &[u8]) -> Result<()>
+    {
+        context_set_property_multiple(self, types::Property::GcmAad, gcm_aad)
+    }
+    /// Sets the GCM tag property.
+    fn set_property_gcm_tag(&self, gcm_tag: &[u8]) -> Result<()>
+    {
+        context_set_property_multiple(self, types::Property::GcmTag, gcm_tag)
+    }
+    /// Sets the CCM AAD property.
+    fn set_property_ccm_aad(&self, ccm_aad: &[u8], input_len: usize) -> Result<()>;
+    // {
+    //     OPERATION_set_input_length(self, input_len)?;
+    //     crypto::context_set_property_multiple(ctx, types::Property::CcmAad, ccm_aad)
+    // }
+    /// Sets the CCM tag property.
+    fn set_property_ccm_tag(&self, ccm_tag: &[u8]) -> Result<()>
+    {
+        context_set_property_multiple(self, types::Property::CcmTag, ccm_tag)
+    }
+}
+
+
+/// Initializes the library. Must be called before any other crypto
+/// function. Should be called once in each thread that uses yaca.
+pub fn initialize() -> Result<()>
+{
+    let r = unsafe {
+        lib::yaca_initialize()
+    };
+    conv::res_c_to_rs(r)
+}
+
+/// Cleans up the library.
+/// Must be called before exiting the thread that called yaca_initialize().
+pub fn cleanup()
+{
+    unsafe {
+        lib::yaca_cleanup()
+    }
+}
+
+/// Safely compares first length bytes of two buffers.
+pub fn memcmp(first: &[u8], second: &[u8], length: usize) -> Result<bool>
+{
+    let min = std::cmp::min(first.len(), second.len());
+    assert!(length <= min);
+    let first = first.as_ptr() as *const c_void;
+    let second = second.as_ptr() as *const c_void;
+    let r = unsafe {
+        lib::yaca_memcmp(first, second, length)
+    };
+    conv::res_c_to_rs_bool(r)
+}
+
+/// Generates random data.
+pub fn random_bytes(length: usize) -> Result<Vec<u8>>
+{
+    let mut v: Vec<u8> = Vec::with_capacity(length);
+    let data = v.as_mut_ptr() as *mut c_char;
+    let r = unsafe {
+        lib::yaca_randomize_bytes(data, length)
+    };
+    conv::res_c_to_rs(r)?;
+    unsafe {
+        v.set_len(length);
+    };
+    Ok(v)
+}
+
+fn context_set_property_single<T, U>(ctx: &T, prop: types::Property,
+                                     value: U) -> Result<()>
+    where T: Context + ?Sized,
+{
+    let ctx = ctx.get_handle();
+    let property = conv::property_rs_to_c(&prop);
+    let value: *const U = &value;
+    let value = value as *const c_void;
```

Comment:
This is extremely easy to misuse because there's no bound on ```U```. This compiles: ```context_set_property_single(self, types::Property::CcmTagLen, vec![0,12])```.

I assume ```context_set_property_single``` is to be used for integral payloads of constant size (```c_int```, ```usize``` and the like), while ```context_set_property_multiple``` variable length is for stuff like the last two args to ```EVP_CipherUpdate```. If I'm right, ```context_set_property_single``` could be made safer by explicitly constraining ```U``` to integral types, either by using sth like https://docs.rs/num/0.2.1/num/trait.Integer.html or by rolling a similar local trait.

Reply:
This is a private function with a sole purpose of wrapping C code. I call it myself internally with proper types. I see no reason to change anything here.

Reply:
Consider this then: this function is not memory safe.

Reply:
Does it have to be? Raw C calls are not as well. Public API is memory safe.

Reply:
As long as it's not marked ```unsafe```, yes.

Reply:
I'll mark it unsafe then.

Reply:
Dangerous but acceptable.

Reply:
Added U: num::Integer as per other thread.